internal/credentials/spiffe.go

```diff
@@ -57,7 +67,7 @@ func SPIFFEIDFromState(state tls.ConnectionState) *url.URL {
 			return nil
 		}
 		// A valid SPIFFE certificate can only have exactly one URI SAN field.
-		if len(state.PeerCertificates[0].URIs) > 1 {
+		if len(cert.URIs) > 1 {
```

Comment:
Can this be checked outside of the for loop? (probably alongside the `nil` check)

Reply:
If we move it outside, the logging can't be moved together because at that time we are not sure the user is using SPIFFE ID yet. In current implementation, we will only log the warning under the condition `uri.Scheme == "spiffe"`(which indicates the user uses SPIFFE ID).

Reply:
I'm not clear if a valid SPIFFE certificate can only have one URI SAN field of any scheme or it can have multiple URI SAN fields, but only one of them can have a scheme of `spiffe`. If its the latter, then the current check does not do the correct thing. It its the former, then you can check at the top before the for loop.